### PR TITLE
Add unit tests for the shamir secret sharing functions.

### PR DESCRIPTION
**Pull request by cbienaime**
Add unit tests for the shamir secret sharing functions.

### CODE REVIEW

File: wrappers/csharp/Helper.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using Devolutions.Cryptography;
+
+namespace DevolutionsCrypto
+{
+    public static class Helper
+    {
+
+        public static IntPtr[] GetArrayReferences(ref byte[][] shares, int nbShares, int secretLength)
```
Comment (pdugre) [inline wrappers/csharp/Helper.cs]: Should not be public

File: wrappers/csharp/Helper.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            IntPtr[] pointers = new IntPtr[(int)nbShares];
+            for (int i = 0; i < handles.Length; i++)
+            {
+                pointers[i] = handles[i].AddrOfPinnedObject();
+            }
+
+            return pointers;
+        }
+
+        public static IntPtr[] InitializeArray(ref byte[][] shares, int nbShares, int secretLength)
```
Comment (pdugre) [inline wrappers/csharp/Helper.cs]: Should not be public

File: wrappers/csharp/Managed.cs [hunk 3]
```diff
@@ -871,6 +873,44 @@ public static string DecryptWithPasswordAsString(string b64data, string password
...
+            var sharesLength = shares[0].Length;
+            var secretLength = (int)Native.JoinSharesSizeNative((UIntPtr)sharesLength);
+            var secret = new byte[secretLength];
+            var tempShares = shares;
+            var sharesTemp = new byte[nbShares][];
+            Array.Copy(shares, sharesTemp, shares.Length);
+            var pointers = Utils.GetArrayReferences(ref shares, nbShares, secretLength);
+            Array.Copy(sharesTemp, shares, sharesTemp.Length);
+            return Native.JoinSharesNative((UIntPtr)nbShares, (UIntPtr)sharesLength, pointers, secret, (UIntPtr)secretLength);
+        }
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: 1. You should check that all the shares have the same length before doing anything and throw an exception is this is not the case.
2. You should free all the pointers before returning.
3. You should handle the native exceptions(see other managed methods for examples.
4. The method should return a byte[], not the secret isn't even returned.
5. Why is there still a "sharesTemp"?

File: wrappers/csharp/Managed.cs [hunk 4]
```diff
@@ -871,6 +873,44 @@ public static string DecryptWithPasswordAsString(string b64data, string password
...
+        /// <param name=""></param>
+        /// <param name=""></param>
+        /// <param name=""></param>
+        /// <returns></returns>
+        public static long GenerateSharedKey(int nbShares, int threshold, int secretLength, ref byte[][] shares)
+        {
+            var pointers = Utils.InitializeArray(ref shares, nbShares, secretLength);
+            var result = Native.GenerateSharedKeyNative((UIntPtr)nbShares, (UIntPtr)threshold, (UIntPtr)secretLength, pointers);
+            return result;
+        }
```
Comment (pdugre) [inline wrappers/csharp/Managed.cs]: 1. You should free the pointers before returning.
2. The method should return a byte[][], not take it as reference.
3. Native errors are not handled properly, see above comment.

File: wrappers/csharp/Utils.cs [hunk 5]
```diff
@@ -372,5 +374,48 @@ public static string Version()
...
+            }
+
+            IntPtr[] pointers = new IntPtr[(int)nbShares];
+            for (int i = 0; i < handles.Length; i++)
+            {
+                pointers[i] = handles[i].AddrOfPinnedObject();
+            }
+
+            return pointers;
+        }
```
Comment (pdugre) [inline wrappers/csharp/Utils.cs]: This is duplicated.

File: wrappers/csharp/unit-tests/TestManaged.cs [hunk 6]
```diff
@@ -340,5 +339,43 @@ public void VerifyPassword()
...
+            var array0 = new byte[] { 0x0d, 0x0c, 0x04, 0x00, 0x00, 0x00, 0x01, 0x00, 0x03, 0x01, 0x80, 0xa4, 0x08, 0x4a, 0xbb, 0xfb, 0x0e, 0x97, 0xdc, 0xa8 };
+            shares[0] = array0;
+            //var array1 = new[] { 0x0d, 0x0c, 0x04, 0x00, 0x00, 0x00, 0x01, 0x00, 0x03, 0x02, 0x84, 0x61, 0x49, 0x6a, 0xbe, 0xc8, 0xe2, 0xf5, 0x71, 0x30 };
+            //shares[1] = array1;
+            var array2 = new byte[] { 0x0d, 0x0c, 0x04, 0x00, 0x00, 0x00, 0x01, 0x00, 0x03, 0x03, 0xd6, 0xa9, 0x57, 0x9d, 0xda, 0xac, 0x41, 0x30, 0x57, 0x76 };
+            shares[1] = array2;
+            //var array3 = new[] { 0x0d, 0x0c, 0x04, 0x00, 0x00, 0x00, 0x01, 0x00, 0x03, 0x04, 0x91, 0x4c, 0x39, 0x9e, 0x83, 0x77, 0x63, 0xc1, 0xca, 0x1a };
+            //shares[3] = array3;
+            var array4 = new byte[] { 0x0d, 0x0c, 0x04, 0x00, 0x00, 0x00, 0x01, 0x00, 0x03, 0x05, 0xc3, 0x84, 0x27, 0x69, 0xe7, 0x13, 0xc0, 0x04, 0xec, 0x5c };
+            shares[2] = array4;
```
Comment (pdugre) [inline wrappers/csharp/unit-tests/TestManaged.cs]: Remove comments

File: wrappers/csharp/unit-tests/TestManaged.cs [hunk 7]
```diff
@@ -340,5 +339,43 @@ public void VerifyPassword()
...
+
+        [TestMethod]
+        public void GenerateSharedKey()
+        {
+            const int nbShares = 5;
+            const int secretLength = 10;
+            const int threshold = 3;
+            var shares = new byte[nbShares][];
+            var result = Managed.GenerateSharedKey(nbShares, threshold, secretLength, ref shares);
+            Assert.AreEqual(0, result);
```
Comment (pdugre) [inline wrappers/csharp/unit-tests/TestManaged.cs]: It should check that the shares generated are not empty, not the result code.

File: wrappers/csharp/unit-tests/TestManaged.cs [hunk 8]
```diff
@@ -340,5 +339,43 @@ public void VerifyPassword()
...
+        }
+
+        [TestMethod]
+        public void JoinShares()
+        {
+            var shares = GetSharesKeys();
+            var result = Managed.JoinShares(shares);
+            Assert.AreEqual(0, result);
```
Comment (pdugre) [inline wrappers/csharp/unit-tests/TestManaged.cs]: It should check that a shared key is created and that the same secret key can be generated using multiple pairs of shares.

**Comment by pdugre** [comment]
Also, the CI will fail until the branch is rebased, see PR https://github.com/Devolutions/devolutions-crypto/pull/113

File: wrappers/csharp/Helper.cs (hunk 1 above)
Comment (cbienaime) [reply wrappers/csharp/Helper.cs]: Files removed

File: wrappers/csharp/Helper.cs (hunk 2 above)
Comment (cbienaime) [reply wrappers/csharp/Helper.cs]: Files removed

File: wrappers/csharp/unit-tests/TestManaged.cs (hunk 6 above)
Comment (cbienaime) [reply wrappers/csharp/unit-tests/TestManaged.cs]: Comment are here for future tests.

**Comment by MathieuMorrissette** [comment]
I will create a new branch. It seems the diff is broken.

**Pull request closed without merging**